include/swift/AST/NameLookup.h

```diff
@@ -27,21 +27,38 @@
 #include "swift/Basic/SourceLoc.h"
 #include "swift/Basic/SourceManager.h"
 #include "llvm/ADT/SmallVector.h"
+#include "swift/AST/TypeRepr.h"
 
 namespace swift {
 class ASTContext;
+class ASTWalker;
 class DeclName;
 class Type;
 class TypeDecl;
 class ValueDecl;
 struct SelfBounds;
 class NominalTypeDecl;
+class Evaluator;
 
 namespace ast_scope {
 class ASTSourceFileScope;
 class ASTScopeImpl;
 } // namespace ast_scope
 
+using CollectedOpaqueReprs = SmallVector<TypeRepr *, 2>;
+
+/// Walk the type representation recursively, collecting any
+/// `OpaqueReturnTypeRepr`s.
+CollectedOpaqueReprs collectOpaqueReturnTypeReprs(TypeRepr *);
+
+/// Walk the type representation recursively, collecting any
+/// `CompositionTypeRepr`s and `IdentTypeRepr`s.
+CollectedOpaqueReprs collectTypeReprs(TypeRepr *);
+
+/// Given a Type Representation, decide if it is a protocol
+bool isProtocol(Evaluator &evaluator, TypeRepr *r, ASTContext &ctx, DeclContext *dc);
```

Comment:
Nit: `ASTContext` has an `evaluator` property, so you can access the evaluator through `ctx.evaluator` rather than passing it separately!

Also, should this be a method on `TypeRepr` instead of a top-level function?

Reply:
`isProtocol` does make more sense as a TypeRepr method, updated!